### PR TITLE
Upgrade Node 14 to Node 20

### PR DESCRIPTION
**Pull request by chris-bateman**
###  Now upgraded to node 20 and webpack 5

Upgrade Node from 14 to 16 and associated packages and WebPack config.

The proposed change is because Node 14 is EOL and some packages are no longer supported.
https://endoflife.date/nodejs
Node 16 is still EOL but I wanted to introduce these changes gradually.

Node is now installed in the Dockerfile using the modern install method.  The old script method is deprecated.
image

Replace ExtractTextPlugin with MiniCssExtractPlugin due to it being deprecated - 
image

I upgraded WebPack slightly to ensure MiniCssExtractPlugin works.

Dockerfile compiled correctly on local system.
Tests completed successfully on local system.
image

ASDC Disclaimer - These changes will benefit the ASDC fork for supporting some custom frontend packages.

### CODE REVIEW

**Comment by chris-bateman** [comment]
A few CSS issues are appearing, so will work on that.

**Comment by pierotofy** [comment]
If we're making the effort to upgrade node/webpack & friends, maybe this should be updated to the latest stable release of node, not 16, unless there's something that's being blocked with staying on 14 (are these the custom frontend packages?)

**Comment by chris-bateman** [comment]
The logic behind moving to 16 was to reduce the blast radius of such a large jump between 14 to 18.  However, 16 is still EOL so best practice would be to move to 18.  I will give it a try and see what the impact is.  It should hopefully be easier* (famous last words) since I have done the move to 16 essentially.

Later versions of Uppy and webpack 5 aren't happy with Node 14 hence the upgrade.

**Comment by chris-bateman** [comment]
@pierotofy in theory Node 18 and webpack 5.
Tests pass locally and in the GH action.
Basic build, login and check various pages and actions appear to work.

Please have a review and let me know if you have any questions or comments.

**Comment by smathermather** [comment]
@chris-bateman -- any reason why 21.04 instead of LTS 22.04? No comments otherwise and a happy to test. Thanks for this!

**Comment by chris-bateman** [comment]
> @chris-bateman -- any reason why 21.04 instead of LTS 22.04? No comments otherwise and a happy to test. Thanks for this!

Didn't touch the docker file Ubuntu version as I didn't want to change too many things.  I was going to consider changing that in another PR.  But overall I do agree a bump there would be good.

**Comment by smathermather** [comment]
Ah, I forgot that was 21 already! Yeah, a bump if possible is always good, but not changing everything at once is useful.

**Comment by smathermather** [comment]
If there's no objection, I'll merge today.

File: Dockerfile [hunk 1]
```diff
@@ -13,9 +14,13 @@ WORKDIR /webodm
...
 RUN printf "deb http://old-releases.ubuntu.com/ubuntu/ hirsute main restricted\ndeb http://old-releases.ubuntu.com/ubuntu/ hirsute-updates main restricted\ndeb http://old-releases.ubuntu.com/ubuntu/ hirsute universe\ndeb http://old-releases.ubuntu.com/ubuntu/ hirsute-updates universe\ndeb http://old-releases.ubuntu.com/ubuntu/ hirsute multiverse\ndeb http://old-releases.ubuntu.com/ubuntu/ hirsute-updates multiverse\ndeb http://old-releases.ubuntu.com/ubuntu/ hirsute-backports main restricted universe multiverse" > /etc/apt/sources.list
 
-# Install Node.js
+# Install Node.js using new Node install method
 RUN apt-get -qq update && apt-get -qq install -y --no-install-recommends wget curl && \
-    wget --no-check-certificate https://deb.nodesource.com/setup_14.x -O /tmp/node.sh && bash /tmp/node.sh && \
+    apt-get install -y ca-certificates gnupg && \
+    mkdir -p /etc/apt/keyrings && \
+    curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg && \
+    NODE_MAJOR=18 && \
```
Comment (pierotofy) [inline Dockerfile:22]: This should be set to 20. There shouldn't be any breaking changes between 18 and 20.

**Comment by pierotofy** [comment]
> Now running as my daily.

Is the version number showing as 3.0.0 in the about section?

**Comment by smathermather** [comment]
Hmm. Nope. Looks like somehow I built wrong.

**Comment by chris-bateman** [comment]
> This breaks the map view for me:
> 
> image
> 
> Probably due to:
> 
> ```javascript
>     fallback: {
>       "buffer": false
>     }
> ```
> 

I will take a look.

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "WebODM",
-  "version": "2.3.0",
+  "version": "3.0.0",
```
[the remark this answers is not in the document]
Comment (chris-bateman) [reply package.json:3]: Yep will fix

File: Dockerfile (hunk 1 above)
Comment (chris-bateman) [reply Dockerfile:22]: I will change to node 20 and review.  However I have seen a few comments for webpack 5 and node 20 having issues.  Node 18 is still supported so there wouldn't be any major issue staying on 18 and waiting for various dependencies to catch up.

**Comment by chris-bateman** [comment]
> This breaks the map view for me:
> 
> image
> 
> Probably due to:
> 
> ```js
>     fallback: {
>       "buffer": false
>     }
> ```

Issue resolved.  Just working on node 20 and seeing what happens there.

**Comment by chris-bateman** [comment]
Latest run failed but just due to a network timeout connecting to Ubuntu.  Happens every now and then.
Built locally just fine on my local system with Node 20.  So just trigger a rebuild and it should be fine.

**Comment by pierotofy** [comment]
Thanks for the additional changes, but this is now breaking shapefile import in the map view. You should be able to drag & drop zipped shapefiles into a project (test files and link to a working instance of WebODM included).

image

[test.zip](https://github.com/OpenDroneMap/WebODM/files/13465356/test.zip)

https://cloud1.webodm.net/public/task/329b1d24-826e-4cbf-a220-2e19be0349e1/map/

You can't just upgrade key libraries and claim "all good". If you've upgraded webpack, test whether you've accidentally broken Javascript related. If you add a polyfill for Buffer, search for anything related to Buffer (in this case, shapefile import, perhaps other stuff), because it might have broken. :pray:

**Comment by smathermather** [comment]
@pierotofy -- do we maintain a list of things to test (recognising this might have been caught without such a list). If not, I can at least go through pull requests and make a list of things that have broken as a starting place and we can build from there.

**Comment by chris-bateman** [comment]
> Thanks for the additional changes, but this is now breaking shapefile import in the map view. You should be able to drag & drop zipped shapefiles into a project (test files and link to a working instance of WebODM included).
> 
> image
> 
> [test.zip](https://github.com/OpenDroneMap/WebODM/files/13465356/test.zip)
> 
> https://cloud1.webodm.net/public/task/329b1d24-826e-4cbf-a220-2e19be0349e1/map/
> 
> If you've upgraded webpack, test whether you've accidentally broken anything Javascript related. If you add a polyfill for Buffer, search for anything related to Buffer (in this case, shapefile import, perhaps other stuff) :pray: 
>  
> 
> 

Thanks for testing.  Comments noted and appreciated. As @smathermather mentions above, a test list would be appreciated.  This could then be expanded into a test suite like Playwright in the future, for example.

I will review the impact of upgrading to WebPack 5 in more detail on a code level.

**Comment by pierotofy** [comment]
* Webpack upgrade: touches every single Javascript output. So everything in the UI/frontend needs to be tested.
* Buffer: any reference (code + libs in node_modules) should be identified and tested.

We have a minimal test suite for the frontend, but it's nowhere near exhaustive (see all .test.jsx files). Manual testing is still required.

**Comment by chris-bateman** [comment]
> Thanks for the additional changes, but this is now breaking shapefile import in the map view. You should be able to drag & drop zipped shapefiles into a project (test files and link to a working instance of WebODM included).
> 
> image
> 
> [test.zip](https://github.com/OpenDroneMap/WebODM/files/13465356/test.zip)
> 
> https://cloud1.webodm.net/public/task/329b1d24-826e-4cbf-a220-2e19be0349e1/map/
> 
> If you've upgraded webpack, test whether you've accidentally broken anything Javascript related. If you add a polyfill for Buffer, search for anything related to Buffer (in this case, shapefile import, perhaps other stuff) 🙏

This particular issue is now fixed.  I will push the commit shortly.
image

Please feel free to test further.  I will continue checking a few things in other sections of the frontend as well.

**Comment by smathermather** [comment]
Do we need more testing on this?

**Comment by chris-bateman** [comment]
> Do we need more testing on this? Happy to run some tests if so.

I am using it for the day to day ASDC workloads without issue but they are probably vastly different from day to day "vanilla" usage.  Maybe release a beta to common power users?

**Comment by smathermather** [comment]
Is there something other than throwing the `--dev` flag that I should do to test it?

**Comment by pierotofy** [comment]
Yes.

`docker build -t opendronemap/webodm_webapp .`

Then `./webodm.sh start --dev`

**Comment by smathermather** [comment]
> Yes.
> 
> `docker build -t opendronemap/webodm_webapp .`
> 
> Then `./webodm.sh start --dev`

Cool. In that case currently running this as a daily (for real this time). Though version still shows as 2.4.0, which is a bit confusing. Perhaps I'm not running it?

Sorry for the confusion: I never touch WebODM, always ODM or NodeODM for testing, so I'm a very old n00b.

**Comment by chris-bateman** [comment]
2.4.0 is right.  I changed it to that from 3.



Sent from my Galaxy



-------- Original message --------
From: Stephen Mather ***@***.***>
Date: 9/12/23 12:27 am (GMT+10:00)
To: OpenDroneMap/WebODM ***@***.***>
Cc: Chris Bateman ***@***.***>, Mention ***@***.***>
Subject: Re: [OpenDroneMap/WebODM] Upgrade Node 14 to Node 20 (PR #1439)


Yes.

docker build -t opendronemap/webodm_webapp .

Then ./webodm.sh start --dev

Cool. In that case currently running this as a daily (for real this time). Though version still shows as 2.4.0, which is a bit confusing. Perhaps I'm not running it?

Sorry for the confusion: I never touch WebODM, always ODM or NodeODM for testing, so I'm a very old n00b.

**Comment by smathermather** [comment]
Still running in production over here without issue, but I've asked @Saijin-Naib to kick the tires a bit too as I haven't had time to do any deeper testing yet.

**Comment by Saijin-Naib** [comment]
Oh, you want me to break stuff? That I can do 🤣

**Comment by smathermather** [comment]
Still running as my daily, so far without incident.

@Saijin-Naib -- any chance to kick the tires?

**Comment by Saijin-Naib** [comment]
Not yet!

**Comment by smathermather** [comment]
Still not seeing any issues running this on the daily. Do we have any remaining reservations about this pull request?

**Comment by pierotofy** [comment]
I think it's good.

Thank you @chris-bateman :pray:

**Comment by chris-bateman** [comment]
Happy new year everyone thanks for taking the time to test this.

**Pull request merged**